tests/v2/fixtures.py

```diff
@@ -0,0 +1,219 @@
+import asyncio
+import copy
+from collections import deque
+from typing import (
+    Any,
+    AsyncGenerator,
+    AsyncIterator,
+    TypeAlias,
+    assert_never,
+)
+
+import msgpack
+import pytest
+from aiochannel import Channel
+from websockets.asyncio.server import ServerConnection, serve
+
+from replit_river.transport_options import TransportOptions, UriAndMetadata
+from replit_river.v2.client import Client
+from tests.v2.datagrams import (
+    FromClient,
+    TestTransport,
+    ToClient,
+    WaitForClosed,
+    parser,
+)
+from tests.v2.interpreter import make_interpreters
+
+# Client -> Server
+ClientToServerChannel: TypeAlias = Channel[Any]
+# Server -> Client
+ServerToClientChannel: TypeAlias = Channel[Any]
+
+
+@pytest.fixture
+async def raw_websocket_meta() -> AsyncIterator[
+    tuple[ServerToClientChannel, ClientToServerChannel, UriAndMetadata[None]]
+]:
+    # server -> client
+    server_to_client = Channel[dict[str, Any]](maxsize=1)
+    # client -> server
+    client_to_server = Channel[dict[str, Any]](maxsize=1)
+
+    # Service websocket connection
+    async def handle(websocket: ServerConnection) -> None:
+        async def emit() -> None:
+            async for msg in client_to_server:
+                await websocket.send(msgpack.packb(msg))
+
+        emit_task = asyncio.create_task(emit())
+        try:
+            while message := await websocket.recv(decode=False):
+                assert isinstance(message, bytes)
+                unpacked = msgpack.unpackb(message, timestamp=3)
+                await server_to_client.put(unpacked)
+        finally:
+            server_to_client.close()
+            client_to_server.close()
+            emit_task.cancel()
+            await emit_task
+        return None
+
+    ipv4_laddr: str | None = None
+    async with serve(handle, "localhost") as server:
+        for sock in server.sockets:
+            if (pair := sock.getsockname())[0] == "127.0.0.1":
+                ipv4_laddr = "ws://%s:%d" % pair
+        serve_forever = asyncio.create_task(server.serve_forever())
+        assert ipv4_laddr
+        yield (
+            server_to_client,
+            client_to_server,
+            UriAndMetadata(uri=ipv4_laddr, metadata=None),
+        )
+
+    serve_forever.cancel()
+
+
+@pytest.fixture
+async def bound_client(
+    raw_websocket_meta: tuple[
+        ClientToServerChannel,
+        ServerToClientChannel,
+        UriAndMetadata[None],
+    ],
+    expected: deque[TestTransport],
+) -> AsyncGenerator[Client[None], None]:
```

Comment:
This is a fly-by-wire'd websocket and `river.v2.Client[None]` that operates entirely based on the passed-in `expected` test script.

There are still some changes that I'd like to implement in the future, namely when you look at the actually written tests you'll see that `ack=0` for all the `FromClient` messages, it's just because the client sends as fast as it can and it gets ahead of the server acks.

To offer some guarantees we just ensure that the `to_client` and `from_client` queues are fully drained as part of the test success criteria.